tests/unit/test_settings.py

```diff
@@ -106,3 +110,104 @@ def test_allow_exist_both_global_and_integration_config(self):
+        self.assertEqual(redis_config.key1, 'value1')
+        self.assertEqual(redis_config.key2, 'value2')
+
+    def test_default_override(self):
+        """
+        Test that a default can be overridden with a user specified value.
+        """
+        redis_config = Config().redis
+        redis_config._add_default('key1', 'value1')
+        redis_config.key1 = 'uservalue1'
+        self.assertEqual(redis_config.key1, 'uservalue1')
+
+    @mock.patch.dict(os.environ, {'DD_REDIS_SERVICE_NAME': 'env-redis'})
+    def test_default_environment_variable(self):
+        """
+        Test that an environment variable overrides the default value.
+        """
+        redis_config = Config().redis
+        redis_config._add_default('service_name', 'my-redis-service')
+        self.assertEqual(redis_config.service_name, 'env-redis')
+
+    @mock.patch.dict(os.environ, {'DATADOG_REDIS_SERVICE_NAME': 'env-redis'})
+    def test_default_environment_variable_legacy(self):
+        """
+        Same as test_default_environment_variable but with the legacy naming.
+        """
+        with mock.patch('warnings.warn') as mock_warn:
+            redis_config = Config().redis
+            redis_config._add_default('service_name', 'my-redis-service')
+            self.assertEqual(redis_config.service_name, 'env-redis')
+            self.assertTrue(mock_warn.call_count > 0)
+
+    @mock.patch.dict(os.environ, {'DD_REDIS_SERVICE_NAME': 'redis_service_name'})
+    def test_no_default_environment_variable(self):
+        """
+        When no default is provided but an environment variable is set
+            An AttributeError should be raised
+        """
+        integration_config = Config().redis
+        with self.assertRaises(AttributeError):
+            self.assertEqual(integration_config.service_name, 'redis_service_name')
+
+    def test_precedence(self):
+        """
+        Test using the config with defaults, environment variables and user
+        overrides.
+        """
+        redis_config = Config().redis
+
+        # Sanity check the default.
+        redis_config._add_default('service_name', 'dd_service_name')
+        self.assertEqual(redis_config.service_name, 'dd_service_name')
+
+        # Check the environment variable.
+        with mock.patch.dict(os.environ, {'DATADOG_REDIS_SERVICE_NAME': 'ude_service_name'}):
+            self.assertEqual(redis_config.service_name, 'ude_service_name')
+
+            # Manual user override while still having the environment variable.
+            redis_config.service_name = 'ud_service_name'
+            self.assertEqual(redis_config.service_name, 'ud_service_name')
+
+            # Check a second manual override.
+            redis_config.service_name = 'ud2_service_name'
+            self.assertEqual(redis_config.service_name, 'ud2_service_name')
+
+    def test_del_key(self):
+        """
+        Test that a user override can be deleted.
+
+        TODO: is this the behaviour we want?
```

Comment:
I'm not sure if it's a valid use-case that someone may want to delete their override at some point. But maybe something worth at least defining with a test case?

Reply:
We can always define `def __delattr__(self, key)` and basically block anything from being deleted from the config.

Seems like an alright approach

Reply:
I opted to leave this undefined for now since we have stopped extending `dict` so `__delattr__` is not defined and `del config[key]` will raise.